core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncSigningStage.java

```diff
@@ -48,9 +63,100 @@ public AsyncSigningStage(HttpClientDependencies dependencies) {
+                } else {
+                    AsyncSignRequest.Builder<T> signRequestBuilder = AsyncSignRequest
+                        .builder(identity)
+                        .request(request)
+                        .payload(context.requestProvider());
+                    authSchemeOption.forEachSignerProperty(signRequestBuilder::putProperty);
+
+                    AsyncSignedRequest signedRequest = signer.signAsync(signRequestBuilder.build());
+                    result = toSdkHttpFullRequest(signedRequest);
+
+                    if (signedRequest.payload().isPresent()) {
+                        context.requestProvider(AsyncRequestBody.fromPublisher(signedRequest.payload().get()));
+                    } else {
+                        // TODO: to confirm. Seems like it is possible that a request with async payload, can go through
+                        //  signAsync and result is no async payload, so better to unset the input async payload than use it
+                        //  as-is if the signer returned null payload?
+                        context.requestProvider(null);
+                    }
+                }
+
+                // TODO: the below updates the result in interceptor context, but if signAsync is called the asyncRequestBody
+                //  could also be different. Should it also be updated in InterceptorContext?
+                //  executionContext.interceptorContext(executionContext.interceptorContext().copy(b -> b.asyncRequestBody(...))).
+                //  This is NOT done in current non-SRA code.
+                updateInterceptorContext(result, context.executionContext());
+                return result;
+            });
+        context.attemptMetricCollector().reportMetric(CoreMetric.SIGNING_DURATION, measuredSign.right());
+        return CompletableFuture.completedFuture(measuredSign.left());
+    }
+
+    private SdkHttpFullRequest toSdkHttpFullRequest(SyncSignedRequest signedRequest) {
+        return toSdkHttpFullRequestBuilder(signedRequest).contentStreamProvider(signedRequest.payload().orElse(null)).build();
+    }
+
+    private SdkHttpFullRequest toSdkHttpFullRequest(AsyncSignedRequest signedRequest) {
+        return toSdkHttpFullRequestBuilder(signedRequest).build();
+    }
+
+    private SdkHttpFullRequest.Builder toSdkHttpFullRequestBuilder(SignedRequest<?> signedRequest) {
+        SdkHttpRequest request = signedRequest.request();
+        return SdkHttpFullRequest.builder()
+                                 .protocol(request.protocol())
+                                 .method(request.method())
+                                 .host(request.host())
+                                 .port(request.port())
+                                 .encodedPath(request.encodedPath())
+                                 .applyMutation(r -> request.forEachHeader(r::putHeader))
+                                 .applyMutation(r -> request.forEachRawQueryParameter(r::putRawQueryParameter));
```

Comment:
This feels inefficient to copy everything. I suppose we have to do this for compatibility with the pipeline, but we should probably move away from it in future refactors.

Can we check to see if signedRequest.request() is already an SdkHttpFullRequest and cast it, if so? I think it should always be true.

Reply:
+1

Good call - added the check to return immediately if already SdkHttpFullRequest